### PR TITLE
[TI] implement CC13X2 CC26X2 OTA feature with ext flash

### PR DESCRIPTION
**Pull request by srickardti**
#### Problem

* The CC13X2_26X2 platform does not have an OTA Image Processor implementation
* The lock-app example does not contain OTA requestor clusters

#### Change overview

 * add OTA clusters to lock app example
 * Implement OTA platform code for cc13x2x7_26x2x7
 * update cc13x2x7_26x2x7 lock app example to initialize the OTA driver
 * update cc13x2x7_26x2x7 applications to an OTA enabled linkerscript
 * create merge tool for the OTA image and the BIM pre-built hexfile

NOTE: Further documentation of OTA process with the TI hardware is required.

#### Testing

* Functionally tested with a RasPi 4B and the `chip-tool`

### CODE REVIEW

File: src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+CHIP_ERROR OTAImageProcessorImpl::SetBlock(ByteSpan & block)
+{
+    if ((block.data() == nullptr) || block.empty())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    // Allocate memory for block data if it has not been done yet
+    if (mBlock.empty())
```
Comment (bzbarsky-apple) [inline src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp:405]: This is buggy.  Please copy one of the non-buggy implementations of SetBlock instead...

@selissia

**Review by bzbarsky-apple — DISMISSED** [review]

File: examples/lock-app/esp32/main/CMakeLists.txt [hunk 2]
```diff
@@ -47,11 +47,13 @@ idf_component_register(INCLUDE_DIRS
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/network-commissioning-old"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/on-off-server"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/operational-credentials-server"
+                     "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/ota-requestor"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/general-commissioning-server"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/diagnostic-logs-server"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/ethernet-network-diagnostics-server"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/thread-network-diagnostics-server"
                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/wifi-network-diagnostics-server"
+                     "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/ota-requestor"
```
Comment (carol-apple) [inline examples/lock-app/esp32/main/CMakeLists.txt:56]: Why is this added twice?

File: src/platform/cc13x2_26x2/OTAImageProcessorImpl.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * Called to release allocated memory for mBlock
+     */
+    CHIP_ERROR ReleaseBlock();
+
+    NVS_Handle mNvsHandle;
+    uint16_t mMetaPage;
+    MutableByteSpan mBlock;
+    // can't this be gotten through the requestor?
```
Comment (carol-apple) [inline src/platform/cc13x2_26x2/OTAImageProcessorImpl.h:62]: This seems like a stale comment?

**Comment by carol-apple** [comment]
Will you have a separate PR to deal with once booted into the new software image, the software version needs to be updated in Configuration Manager as well as calling `OTARequestor::NotifyUpdateApplied` with the new version?

File: examples/lock-app/esp32/main/CMakeLists.txt (hunk 2 above)
Comment (srickardti) [reply examples/lock-app/esp32/main/CMakeLists.txt:56]: I was copying the lighting-app CMakeLists.txt

That didn't fix the automated build, so I will have to look into it further. If someone with more experience on these platforms can help I would appreciate it.

File: src/platform/cc13x2_26x2/OTAImageProcessorImpl.h (hunk 3 above)
Comment (srickardti) [reply src/platform/cc13x2_26x2/OTAImageProcessorImpl.h]: removed

**Comment by srickardti** [comment]
> Will you have a separate PR to deal with once booted into the new software image, the software version needs to be updated in Configuration Manager as well as calling `OTARequestor::NotifyUpdateApplied` with the new version?

I don't see any other embedded platform notify of an update. I can write the software version to internal non-volatile storage, but I'm not sure how to persist the requestor. If there is an example of that I would be glad to copy it.

**Comment by carol-apple** [comment]
> > Will you have a separate PR to deal with once booted into the new software image, the software version needs to be updated in Configuration Manager as well as calling `OTARequestor::NotifyUpdateApplied` with the new version?
> 
> I don't see any other embedded platform notify of an update. I can write the software version to internal non-volatile storage, but I'm not sure how to persist the requestor. If there is an example of that I would be glad to copy it.

`NotifyUpdateApplied` was a recent add and I've been communicating with everyone that it should be called by each platform developer once running the new software image. The Requestor will use Configuration Manager as the source of truth so that's where it needs to make sure the software version is updated.

**Comment by srickardti** [comment]
> `NotifyUpdateApplied` was a recent add and I've been communicating with everyone that it should be called by each platform developer once running the new software image. The Requestor will use Configuration Manager as the source of truth so that's where it needs to make sure the software version is updated.

I'd be willing to add this, but I'm not quite sure where to do it. The only reference I can find is the Linux implementation, and I'm not sure how that can be applied here.

The values in the configuration manager will be preserved across the reset and subsequent image update. That means the device can connect back to the Thread network and will have the Fabric info, but the OTA requestor does not seem to persist any of the OTA data. Specifically the provider node and endpoint id and the version sent with the OTA protocol is not persisted through the configuration manager. 

This seems like a pretty common case for embedded devices that have to reset to apply new firmware or even non-embedded devices that would want to restart their daemon processes to apply an update. There might be a cleaner way to implement this within the OTA requestor itself, but again I'm not sure where the best place would be.

**Comment by carol-apple** [comment]
> > `NotifyUpdateApplied` was a recent add and I've been communicating with everyone that it should be called by each platform developer once running the new software image. The Requestor will use Configuration Manager as the source of truth so that's where it needs to make sure the software version is updated.
> 
> I'd be willing to add this, but I'm not quite sure where to do it. The only reference I can find is the Linux implementation, and I'm not sure how that can be applied here.

Yeah, the Linux OTA Requestor does not persist any data as of yet. 

> 
> The values in the configuration manager will be preserved across the reset and subsequent image update. That means the device can connect back to the Thread network and will have the Fabric info, but the OTA requestor does not seem to persist any of the OTA data. Specifically the provider node and endpoint id and the version sent with the OTA protocol is not persisted through the configuration manager.

I am currently working on supporting DefaultOTAProviders attribute. This attribute will contain provider node ID, fabric index, and endpoint. Once this attribute has been supported, we will add persisting that attribute.

With that said, software version should still come from Configuration Manager but for Linux, this is not something that can be set (in Configuration Manager). I'll let @shubhamdp comment on how ESP32 plans on updating software version.

> 
> This seems like a pretty common case for embedded devices that have to reset to apply new firmware or even non-embedded devices that would want to restart their daemon processes to apply an update. There might be a cleaner way to implement this within the OTA requestor itself, but again I'm not sure where the best place would be.

**Comment by shubhamdp** [comment]
@srickardti, AFAIK, Configuration manger reads software version by reading [CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION](https://github.com/project-chip/connectedhomeip/blob/master/src/include/platform/internal/GenericConfigurationManagerImpl.h#L165-L169) macro. Basically you'll need a way to define this macro. In PR #14693 I have tried making the software version configurable, which can be set from application.

**Comment by tima-q** [comment]
@srickardti - are you pushing this PR forward ? We're looking to add OTA to our lock-app example as well, and this PR already contains the common changes.
@carol-apple The changes around the 'NotifyUpdateApplied' API - can that be captured in a follow-up PR ? If I understand correctly it's actually a generic function for all platforms, to be hooked to the Software version being different then what's inside the Configuration Manager ? So the flow can be the same for all ? 
- Init 
- Check version in define/const <> Configuration Mem contents 
-- if different, trigger NotifyUpdateApplied 
- all cases - write const to Configuration Mem

**Comment by woody-apple** [comment]
Fast tracking, this has been in review for a long time.

**Review by woody-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
